ipq40xx: add basic 6.12 kernel support

DragonBluep opened the pull request:
### Todo:
- [x] Waiting for the prerequisites to be merged.
https://github.com/openwrt/openwrt/pull/18637
https://github.com/openwrt/openwrt/pull/18744
https://github.com/openwrt/openwrt/pull/19172
- [x] Self-review Kconfig symbol changes.

Tested on SKSpruce WIA3300-20 (ipq40xx/generic).

DragonBluep committed:
ipq40xx: refresh kernel symbol configs

This is a preparation for introducing the 6.12 kernel support.
All configs are automatically refreshed. In theory, they will
generate the same .config files in the kernel build directory
as before.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: gcc-ipq4019: mark qcom_find_freq_close() function as static

This function is only used by the current source file gcc-ipq4019.c.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: refresh 6.12 kernel patches

Remove upstreamed:
004-v6.7-firmware-qcom_scm-disable-SDI-if-required.patch [1]
709-ARM-dts-qcom-ipq4019-add-QCA8075-PHY-Package-nodes.patch [2]

Manually rebased:
422-firmware-qcom-scm-fix-SCM-cold-boot-address.patch
701-net-dsa-add-out-of-band-tagging-protocol.patch
850-soc-add-qualcomm-syscon.patch
900-PCI-qcom-add-hack-compatible-for-ipq4019-Lantiq-DSL.patch
910-Revert-firmware-qcom_scm-Clear-download-bit-during-r.patch

[1] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=ff4aa3bc98258a240b9bbab53fd8d2fb8184c485
[2] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=2338f4315f16b937e924ff679b91bb8c0ab53f25
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: ipqess: remove explicit phylink_generic_validate() references

The phylink_generic_validate() can be automatically called if
phylink_validate method was not given in the switch operations.

Link: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.6.y&id=5938bce4b6e2146d5194badd7ecf0acc6bf877fc
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: rename patchset and kernel configs to 6.12

This is a preparation for 6.12 kernel support. It can help us
track the patches and Kconfig history by using the Git tool.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: tag_oob/ipqess: update DSA APIs

In the new 6.12 LTS kernel, master/slave keywords were replaced
by conduit/user.

Link: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=6ca80638b90cec66547011ee1ef79e534589989a
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: move arch files to 6.12 directory

This is a preparation for 6.12 kernel support. It can help us
track the device tree files history by using the Git tool.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: qca8k-ipq4019: add missing header platform_device.h

The prototype of `devm_platform_ioremap_resource_byname()` was defined
in `linux/platform_device.h`.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: restore kernel 6.6 arch files

Copy arch directory from 6.12 kernel to restore the default
6.6 device tree files.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: refresh 6.12 kernel config files

Manually selected symbols:

- Enable ARCH_QCOM_RESERVE_SMEM
  Reserve SMEM at the beginning of RAM

- Enable QCOM_TZMEM_MODE_GENERIC
  TrustZone interface memory allocator mode

- Disable QCOM_QSEECOM
  Qualcomm QSEECOM interface driver

- Disable IPQ_NSSCC_QCA8K
  QCA8K(QCA8386 or QCA8084) NSS Clock Controller

- Disable MSM_IOMMU
  MSM IOMMU Support

- Disable INTERCONNECT_QCOM
  Qualcomm Network-on-Chip interconnect drivers

All other symbols are automatically refreshed by
`make kernel_oldconfig CONFIG_TARGET=target` and
`make kernel_oldconfig CONFIG_TARGET=subtarget`.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: restore kernel 6.6 config files and patches

Copy patches and kernel configs from 6.12 kernel to restore the
default 6.6 kernel support files.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: enable 6.12 testing kernel

The 6.12 testing kernel for ipq40xx target is ready now.

Tested on SKSpruce WIA3300-20.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: add PCIe bridge node reference label for ipq4019.dtsi

Add a bridge node label so that we can insert PCIe peripheral nodes.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

mrkiko commented:
Thanks for the work!

jmv2009 commented:
Building for Fritz Repeater 1200, with usb support not disabled during linux kernel build:

make[5]: Entering directory '/home/kali/zram1/openwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-6.12.26'
  CC      drivers/usb/host/pci-quirks.o
drivers/usb/host/pci-quirks.c:621:6: error: redefinition of 'usb_asmedia_modifyflowcontrol'
  621 | void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev)
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from drivers/usb/host/pci-quirks.c:22:
drivers/usb/host/pci-quirks.h:49:20: note: previous definition of 'usb_asmedia_modifyflowcontrol' with type 'void(struct pci_dev *)'
   49 | static inline void usb_asmedia_modifyflowcontrol(struct pci_dev *pdev) {}
      |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[10]: *** [scripts/Makefile.build:229: drivers/usb/host/pci-quirks.o] Error 1

DragonBluep commented:
> Here's a fix for the dsl setup: [0001-vrx518_ep-fix-compilation-error-with-kernel-6.12.patch.txt](https://github.com/user-attachments/files/20065331/0001-vrx518_ep-fix-compilation-error-with-kernel-6.12.patch.txt)
> 
> Feel free to add it here, it's only compile time tested so far

@dhewg Thanks! I picked your patch into https://github.com/openwrt/openwrt/pull/18744 with some modifications so that we can merge it earlier.

DragonBluep commented:
> Building for Fritz Repeater 1200, with usb support not disabled during linux kernel build, same as in the test builds.
> 
> Without usb it compiles just fine, so looks like some minor tweaking is in order.

@jmv2009 You need to rebase/merge PR https://github.com/openwrt/openwrt/pull/18637 to fix the USB build error. Thanks for your test.

jmv2009 commented:
The Fritz Repeater 1200 doesn't actually have any USB, so for me it already is a go. I don't know why it was trying to get compiled in.

DragonBluep committed:
ipq40xx: dts: fix dtc warnings for MikroTik hAP ac3 LTE6

Correct the mtd partition reg property size to fix the following
dtc warnings:

qcom-ipq4019-hap-ac3-lte6-kit.dts:202.5-25: Warning (reg_format): /soc/spi@78b5000/flash@0/partitions/partition@0:reg: property has invalid length (8 bytes) (#address-cells == 2, #size-cells == 1)
qcom-ipq4019-hap-ac3-lte6-kit.dts:209.5-29: Warning (reg_format): /soc/spi@78b5000/flash@0/partitions/partition@80000:reg: property has invalid length (8 bytes) (#address-cells == 2, #size-cells == 1)
qcom-ipq4019-hap-ac3-lte6-kit.dts:226.5-31: Warning (reg_format): /soc/spi@78b5000/flash@0/partitions/partition@110000:reg: property has invalid length (8 bytes) (#address-cells == 2, #size-cells == 1)

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: dts: fix led-controller node name for Netgear LBR20

The node name should include the reg property value. Fix warning:

qcom-ipq4019-lbr20.dts:415.17-449.4: Warning (i2c_bus_reg): /soc/i2c@78b7000/led-controller: I2C bus unit address format error, expected "27"

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jmv2009 commented:
This should be rebased, now the prerequisites have been merged.

DragonBluep committed:
ipq40xx: dts: compatibility changes for 6.12 kernel

- Remove duplicate PCIe bridge. A default bridge node has been
  added upstream.
- Do not refer to the qca807x node. This node label has been
  removed. We can disable the qca807x phy in mdio node.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: dts: correct PCIe device node name

According to the dtc source code, the PCIe device node unitname
needs to follow the following naming rules:

```
reg = fdt32_to_cpu(cells[0]);
dev = (reg & 0xf800) >> 11;
func = (reg & 0x700) >> 8;
snprintf(unitname, sizeof(unitname), "%x,%x", dev, func);
```

These devices' reg cell[0] is equal to 0x10000, hence the correct
node unitname should be "0,0". This patch fixes the following dtc
warnings for 6.12 kernel:

qcom-ipq4019-map-ac2200.dts:211.11-216.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-rt-ac42u.dts:315.11-319.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-fritzbox-7530.dts:318.10-321.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/dsl@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-fritzrepeater-3000.dts:239.11-245.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-gl-b2200.dts:340.11-348.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-ea8300.dts:97.11-102.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-mr8300.dts:83.11-88.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-whw03v2.dts:234.11-241.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4029-insect-common.dtsi:282.11-287.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4029-insect-common.dtsi:282.11-287.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-lbr20.dts:488.11-495.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-orbi.dtsi:317.11-322.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-a62.dts:218.11-225.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-pa2200.dts:198.11-205.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-mf18a.dts:478.11-484.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"
qcom-ipq4019-mf289f.dts:429.11-435.4: Warning (pci_device_reg): /soc/pcie@40000000/pcie@0/wifi@1,0: PCI unit address format error, expected "0,0"

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

DragonBluep committed:
ipq40xx: dts: fix incorrect PCIe bus number

On the ipq40xx platform, the bus range is 1 - 255. Therefore, bus
number 0 is clearly incorrect. Change the bus number to 1, just
like other ipq40xx devices. This patch fixes the following dtc
warnings on 6.12 kernel:

qcom-ipq4019.dtsi:476.5-29: Warning (pci_device_bus_num): /soc/pcie@40000000/pcie@0/wifi@0,0:bus-range: PCI bus number 0 out of range, expected (1 - 255)
qcom-ipq4019.dtsi:476.5-29: Warning (pci_device_bus_num): /soc/pcie@40000000/pcie@0/wifi@0,0:bus-range: PCI bus number 0 out of range, expected (1 - 255)

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

xdarklight commented:
Thank you @DragonBluep, @dhewg and all others involved!

This gets my:
`Tested-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>`
using a Fritz!Box 7520 (re-badged 7530), including:
```
# dmesg | grep showtime
[  355.857878] vrx518_tc:ptm_showtime_enter: Line[0]: show time enter!!
```

GoetzGoerisch reviewed (approved):
run-tested: ipq40xx/chromium (Google WiFi (Gale))

Thank you @DragonBluep

DragonBluep committed:
ipq40xx: migrate wifi configuration device paths for 6.12 kernel

The device tree PCIe host node name has been changed in the new
6.12 kernel[1]. Hence we have to update the wifi device path to
make sure it can work properly.

This script is based on:
target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate

[1] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.12.y&id=07299ba2e7d98045e6b522f7c5b97f402b15bc82
Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/18725
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

GoetzGoerisch commented:
running also the latest commit (rebased on main) since a couple of days.
@namiltd what does it need to move this forward?

vochong commented:
@DragonBluep

I built an image for MR33 with K6.12.40.   Most things worked well except some weird errors as shown below.

```
[   49.319138] ath10k_ahb a800000.wifi: pdev param 0 not supported by firmware
[   49.324737] ath10k_ahb a800000.wifi: failed to enable peer stats info: -95
[   49.331581] phy2-sta0: associated
**[   69.122483] Alignment trap: not handling instruction edc40b02 at [<000191d4>]
[   69.122558] 8<--- cut here ---
[   69.129393] Unhandled fault: alignment exception (0x801) at 0xb6568721
[   69.131592] [b6568721] *pgd=8992c835, *pte=8b99675f, *ppte=8b996c7f**

```

```
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
Sat Aug  2 13:42:03 2025 daemon.err mgmtd[6919]: [VNSS1-Q9XX7] XXX txn in progress, retry init
```
